packages/imperative/src/config/src/ProfileInfo.ts

```diff
@@ -1225,6 +1234,312 @@ export class ProfileInfo {
+            const type = typeWithPath.split(":").pop();
+            if (type == null) {
+                continue;
+            }
+
+            finalSchema[type] = schema;
+        }
+
+        let schemaEntries = Object.entries(finalSchema);
+        if (sources?.length > 0) {
+            schemaEntries = schemaEntries.filter(([typ, sch]) => {
+                if (!(typ in this.mExtendersJson.profileTypes)) {
+                    return false;
+                }
+
+                // If a list of sources were provided, ensure the type is contributed by at least one of these sources
+                if (sources.some((val) => this.mExtendersJson.profileTypes[typ].from.includes(val))) {
+                    return true;
+                }
+
+                return false;
+            });
+        }
+
+        return ConfigSchema.buildSchema(schemaEntries.map(([type, schema]) => ({
+            type,
+            schema
+        })));
+    }
+
+    /**
+     * @param [sources] (optional) Only include available types from the given list of sources
+     * @returns a list of all available profile types
+    */
+    public getProfileTypes(sources?: string[]): string[] {
+        const filteredBySource = sources?.length > 0;
+        const profileTypes = new Set<string>();
+        for (const layer of this.getTeamConfig().mLayers) {
+            if (layer.properties.$schema == null) continue;
+            const schemaUri = new url.URL(layer.properties.$schema, url.pathToFileURL(layer.path));
+            const schemaPath = url.fileURLToPath(schemaUri);
+            if (!fs.existsSync(schemaPath)) continue;
+
+            const profileTypesInLayer = [...this.mProfileSchemaCache.keys()].filter((key) => key.includes(`${layer.path}:`));
+            for (const typeWithPath of profileTypesInLayer) {
+                const type = typeWithPath.split(":").pop();
+                if (type == null) {
+                    continue;
+                }
+
+                profileTypes.add(type);
+            }
+        }
+
+        // Include all profile types from extenders.json if we are not filtering by source
+        if (filteredBySource) {
+            return [...profileTypes].filter((t) => {
+                if (!(t in this.mExtendersJson.profileTypes)) {
+                    return false;
+                }
+
+                return this.mExtendersJson.profileTypes[t].from.some((src) => sources.includes(src));
+            }).sort((a, b) => a.localeCompare(b));
+        }
+
+        return [...profileTypes].sort((a, b) => a.localeCompare(b));
+    }
+
+    /**
+     * Returns the schema object belonging to the specified profile type.
+     *
+     * @param {string} profileType The profile type to retrieve the schema from
+     * @returns {IProfileSchema} The schema object provided by the specified profile type
+     */
+    public getSchemaForType(profileType: string): IProfileSchema {
+        let finalSchema: IProfileSchema = undefined;
```

Comment:
The previous function return an empty array, and this one undefined for invalid profileTypes.
I'm happy with the way that we are handling this scenario in both functions.

However, if we decide to throw an error in the `getProfileTypes`, then we probably should do the same here 😋 

I'm not advising we do, though. Just something to consider 😋